### PR TITLE
Implement custom sections

### PR DESCRIPTION
**Pull request by bjorn3**
See https://github.com/bjorn3/rustc_codegen_cranelift/pull/435#issuecomment-477429904

This hasn't been tested with mach-O yet.

### CODE REVIEW

**Comment by bjorn3** [comment]
Pushed a commit to bjorn3/rustc_codegen_cranelift#435 to use this, so mach-O gets tested on travis.

File: src/artifact/decl.rs [hunk 1]
```diff
@@ -170,7 +196,7 @@ pub enum DefinedDecl {
     /// A data object defined in this artifact
     Data(DataDecl),
     /// A DWARF debug section defined in this artifact
-    DebugSection(DebugSectionDecl),
+    Section(SectionDecl),
```
Comment (philipc) [inline src/artifact/decl.rs]: nit: outdated comment

File: src/mach.rs [hunk 2]
```diff
@@ -787,13 +799,20 @@ fn build_relocations(segment: &mut SegmentBuilder, artifact: &Artifact, symtab:
                 // NB: we currently deduce the meaning of our relocation from from decls -> to decl relocations
                 // e.g., global static data references, are constructed from Data -> Data links
                 match (link.from.decl, link.to.decl) {
-                    (Decl::Defined(DefinedDecl::DebugSection { .. }), _) => {
+                    (Decl::Defined(DefinedDecl::Section(s)), _)
+                        if s.kind() == SectionKind::Debug => {
```
Comment (philipc) [inline src/mach.rs]: The indentation here is weird. Need to run rustfmt?

Comment (bjorn3) [reply src/mach.rs]: Yes :)

**Comment by bjorn3** [comment]
The mach-O version might not work correctly, or I am using `object` wrong.

https://travis-ci.org/bjorn3/rustc_codegen_cranelift/jobs/512168839#L375

```
 WARN 2019-03-27T19:12:27Z: rustc_metadata::locator: no metadata found: no .rustc section
```

I will investigate tomorrow.

**Comment by bjorn3** [comment]
Added an objdump invocation to bjorn3/rustc_codegen_cranelift#435:

```
target/out/libmini_core.dylib:	file format Mach-O 64-bit x86-64
Sections:
Idx Name          Size      Address          Type
  0 __text        0000061f 0000000000000990 TEXT 
  1 __unwind_info 00000048 0000000000000fb0 DATA 
[BUILD] example
 WARN 2019-03-29T18:53:28Z: rustc_metadata::locator: no metadata found: no .rustc section
 WARN 2019-03-29T18:53:28Z: rustc_metadata::locator: no metadata found: no .rustc section
```

Conclusion: `.rustc` doesn't get emitted at all.

**Comment by philipc** [comment]
Added this code to `src/bin/main.rs` (might be good to include something like this in the PR):
```rust
    obj.declare(".faerie", Decl::section(SectionKind::Data))?;
    obj.define(".faerie", b"faerie".to_vec())?;
```

Then:
```
$ target/debug/prototype --mach mach
$ llvm-objdump -section-headers mach

mach:	file format Mach-O 64-bit x86-64

Sections:
Idx Name          Size      Address          Type
  0 __text        00000053 0000000000000000 TEXT 
  1 __data        0000000c 0000000000000053 DATA 
  2 __cstring     00000017 000000000000005f DATA 
  3 __faerie      00000006 0000000000000076 DATA 
```

**Comment by bjorn3** [comment]
It seems like the section is emitted correctly, but when linking the final dylib it is omitted.

**Comment by bjorn3** [comment]
> might be good to include something like this in the PR)

Done

**Comment by m4b** [comment]
@bjorn3 I'm seeing the section emitted correctly on mach, along with its data "some data".

E.g.:
```
target/debug/prototype --deadbeef --mach deadbeef.mach.o
target/debug/prototype --mach mach.o
target/debug/prototype -l mach --mach mach.o deadbeef.mach.o

./mach
deadbeef: 0xdeadbef0 - 0xcafebabe

strings mach
deadbeef: 0x%x - 0x%x
some data

bingrep --hex mach | grep fear
00000370  00 00 00 00 00 00 00 00 5F 5F 66 61 65 72 69 65  | ........__faerie |

  Segment   # Sections  
  __DATA    4           
         Idx   Name              Addr           Size    Offset    Align   Reloff   Nrelocs   Flags  
         0     __nl_symbol_ptr   0x100001000    0x10    0x1000    3       0x0      0         0x6    
         1     __la_symbol_ptr   0x100001010    0x8     0x1010    3       0x0      0         0x7    
         2     __data            0x100001018    0x17    0x1018    3       0x0      0         0x0    
         3     __faerie          0x100001030    0x9     0x1030    1       0x0      0         0x0  
```

unless I'm missing something? It would be easy to add a test for this, we could emit the object file with goblin, then parse the object, check to see the data is in the section.

Anyway should we merge?

**Comment by bjorn3** [comment]
I think so. The problem with cg_clif is probably caused by different linker invocations than cg_llvm uses.

**Comment by bjorn3** [comment]
Ping @m4b

**Comment by m4b** [comment]
gosh I'm really sorry @bjorn3 merging this now...

**Comment by bjorn3** [comment]
No problem. The PR I need this for is still blocked on a rustc PR anyway.

**Pull request merged**